solidity/contracts/adapters/UniswapV3Adapter.sol

```diff
@@ -142,35 +142,46 @@ contract UniswapV3Adapter is AccessControl, SimpleOracle, IUniswapV3Adapter {
     bytes32 _pairKey = _keyForPair(_tokenA, _tokenB);
     if (_isPairDenylisted[_pairKey]) revert PairCannotBeSupported(_tokenA, _tokenB);
 
-    uint16 _cardinality = uint16((uint32(period) * cardinalityPerMinute) / 60) + 1;
-    address[] memory _pools = UNISWAP_V3_ORACLE.prepareAllAvailablePoolsWithCardinality(_tokenA, _tokenB, _cardinality);
+    address[] memory _pools = _getAllPoolsSortedByLiquidity(_tokenA, _tokenB);
     if (_pools.length == 0) revert PairCannotBeSupported(_tokenA, _tokenB);
 
+    // Load to mem to avoid multiple storage reads
     address[] storage _storagePools = _poolsForPair[_pairKey];
-    uint256 _currentPools = _storagePools.length;
-    uint256 _min = _currentPools < _pools.length ? _currentPools : _pools.length;
-
-    uint256 i;
-    for (; i < _min; i++) {
-      // Rewrite storage
-      _storagePools[i] = _pools[i];
-    }
-    if (_currentPools < _pools.length) {
-      // If have more pools than before, then push
-      for (; i < _pools.length; i++) {
-        _storagePools.push(_pools[i]);
+    uint256 _poolsPreviouslyInStorage = _storagePools.length;
+    uint32 _gasCostPerCardinality = gasPerCardinality;
+
+    uint16 _targetCardinality = uint16((uint32(period) * cardinalityPerMinute) / 60) + 1;
+    uint256 _preparedPools;
+    for (uint256 i; i < _pools.length; i++) {
+      address _pool = _pools[i];
+      (, , , , uint16 _currentCardinality, , ) = IUniswapV3Pool(_pool).slot0();
+      if (_currentCardinality < _targetCardinality) {
+        if (uint32(_targetCardinality - _currentCardinality) * _gasCostPerCardinality > gasleft()) {
+          continue;
+        }
+        IUniswapV3Pool(_pool).increaseObservationCardinalityNext(_targetCardinality);
```

Comment:
Oof. The only thing that would change it, it would be a change in the gas cost parameters on the EVM, which... Can happen eventually, but we can give ourselves a little bit of wiggle room. Not completely sure tbh.